sdks/python/apache_beam/utils/multi_process_shared.py

```diff
@@ -0,0 +1,153 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Implements a shared object that spans processes.
+
+This object will be instanciated once per VM and methods will be invoked
+on it via rpc.
+"""
+# pytype: skip-file
+
+import logging
+import os
+import multiprocessing.managers
+import pickle
+import time
+import threading
+import uuid
+
+from typing import Any
+from typing import Callable
+from typing import Generic
+from typing import Optional
+from typing import TypeVar
+
+import fasteners
+
+T = TypeVar('T')
+
+_singletons = {}
+
+
+def _create_singleton(constructor, tag):
+  _singletons[tag] = constructor()
+
+
+def _get_singleton(tag):
+  return _singletons[tag]
+
+
+class _SingletonManager(multiprocessing.managers.BaseManager):
+  pass
+
+
+_SingletonManager.register('get_singleton', callable=_get_singleton)
+
+_process_local_lock = threading.Lock()
+
+
+class MultiProcessShared(Generic[T]):
+  """MultiProcessShared is used to share a single object across processes.
+
+  For example, one could have the class::
+
+    class MyExpensiveObject(object):
+      def __init__(self, args):
+        [expensive initialization and memory allocation]
+
+      def method(self, arg):
+        ...
+
+  One could share a single instance of this class by wrapping it as::
+
+    my_expensive_object = MultiProcessShared(lambda: MyExpensiveObject(...))
+
+  which could then be invoked as::
+
+    my_expensive_object.method(arg)
+
+
+  Args:
+    constructor: function that initialises / constructs the object if not
+      present in the cache. This function should take no arguments. It should
+      return an initialised object, or raise an exception if the object could
+      not be initialised / constructed.
+    tag: an optional indentifier to store with the cached object. If multiple
+      MultiProcessShared instances are created with the same tag, they will all
+      share the same proxied object.
+    path: a temporary path in which to create the inter-process lock
+    always_proxy: whether to direct all calls through the proxy, rather than
+      call the object directly for the process that created it
+  """
+  def __init__(
+      self,
+      constructor: Callable[[], T],
+      tag: Optional[Any] = None,
+      *,
+      path: str = '/tmp',
```

Comment:
nit: default to tempfile.gettempdir instead of /tmp ?

Reply:
Yes. Done.